src/Helpers.php

```diff
@@ -590,24 +590,129 @@ public static function record_warnings($errno, $errstr, $errfile, $errline)
     }
 
     /**
-     * @param $c
-     * @return bool|string
+     * Get Unicode code point of character
+     *
+     * Shim for use on systems running PHP < 7.2
+     *
+     * @param string $c
+     * @param string $encoding
+     * @return int|false
      */
-    public static function unichr($c)
+    public static function uniord(string $c, string $encoding = null)
     {
+        static $cache = [];
```

Comment:
I wondered if the cache does much here really … So doing some rudimentary benchmarking of different variants (1 million iterations calling the functions with a fixed codepoint), I get results like the following on my machine (numbers are comparable on re-runs):

`mb_ord` variant | time
----------------------- | ------
`mb_ord` direct | 37.32 ms
`Helpers::uniord` | 79.99 ms
`Helpers::uniord` using `\array_key_exists` | 63.56 ms
`Helpers::uniord` without cache | 101.53 ms
`Helpers::uniord` just wrapping `mb_ord` | 59.54 ms

`mb_chr` variant | time
----------------------- | ------
`mb_chr` direct | 40.18 ms
`Helpers::unichr` | 77.17 ms
`Helpers::unichr` using `\array_key_exists` | 57.20 ms
`Helpers::unichr` without cache | 99.42 ms
`Helpers::unichr` just wrapping `mb_chr` | 67.29 ms

Note, the second `Helpers::unichr` variant just adds a leading backslash to `array_key_exists($c, $cache)` (see https://tideways.com/profiler/blog/compiler-optimized-php-functions for background).

So it seems, yes the cache speeds the operation up, but extra function calls and checks quickly eat up any benefits. Using the native functions directly seems to be significantly faster still.

Reply:
Did a re-run using random codepoints; seems the cache can start to slow things down if it gets too large, e.g. using `rand(0, 2000)` as input for `Helpers::unichr` (haven’t thought about whether that is realistic at all):

`mb_chr` variant | time
---------------------- | ------
`mb_chr` direct | 86.87 ms
`Helpers::unichr` | 201.40 ms
`Helpers::unichr` using `\array_key_exists` | 186.53 ms
`Helpers::unichr` without cache | 144.91 ms
`Helpers::unichr` just wrapping `mb_chr` | 107.66 ms

I wonder if these differences would even make any noticeable impact when looking at the overall render time. So my hunch would be to just avoid any caching here, not worth the effort.

Reply:
Note: All tests done on PHP 8.2.

Reply:
I think one of the issues I was fighting against in my performance testing was that I had XDebug enabled and forgot the fairly significant impact it can have on performance. I did all the optimization work prior to generating actual performance numbers over multiple runs. Whoops. It definitely helped find some things that could impact performance, but probably not in a significant way except on slower systems.

With the discussion of bumping the minimum version I think it's safe to say removing this shim entirely is worthwhile.

Reply:
> I think one of the issues I was fighting against in my performance testing was that I had XDebug enabled and forgot the fairly significant impact it can have on performance.

Oh, yeah, XDebug definitely hit me, too. Switched to PHP 7.1 for comparison and wondered why everything was suddenly ten times as fast …

Reply:
FYI the worst-case scenario for the caching logic appears to be https://eclecticgeek.com/dompdf/core_tests/encoding_utf-8_all.html. It's taking minutes to render the document rather than seconds. Caching hurts in this case because most characters are not reused so a lot of work is done to maintain a cache that is unnecessary. So I think you're right, caching overall is not helpful.